### PR TITLE
[IGA] Add refinement modeler

### PR DESCRIPTION
**Pull request by tteschemacher**
**Description**
This PR adds the refinement modeler which makes use of #8360 . Yet, only one avenue of refinements is supported, however, more can be attached easily.

### CODE REVIEW

File: applications/IgaApplication/custom_modelers/refinement_modeler.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+            << "Missing \"parameters\" in refinements block.\n"
+            << rParameters << std::endl;
+        KRATOS_ERROR_IF_NOT(rParameters.Has("geometry_type"))
+            << "Missing \"geometry_type\".\n"
+            << rParameters << std::endl;
+        if (rParameters["geometry_type"].GetString() == "NurbsSurface") {
+            for (IndexType n = 0; n < geometry_list.size(); ++n) {
+                auto p_nurbs_surface = (geometry_list[n].size() > 0)
+                    ? dynamic_pointer_cast<NurbsSurfaceGeometry<3, PointerVector<Node<3>>>>(geometry_list(n))
+                    : dynamic_pointer_cast<NurbsSurfaceGeometry<3, PointerVector<Node<3>>>>(geometry_list(n)->pGetGeometryPart(GeometryType::BACKGROUND_GEOMETRY_INDEX));
```
Comment (manuelmessmer) [inline applications/IgaApplication/custom_modelers/refinement_modeler.cpp:74]: Is this one for nurbs_curve_in_nurbs_surface?

File: applications/IgaApplication/custom_modelers/refinement_modeler.cpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        GeometriesArrayType geometry_list;
+        GetGeometryList(geometry_list, r_model_part, rParameters);
+
+        KRATOS_ERROR_IF_NOT(rParameters.Has("parameters"))
+            << "Missing \"parameters\" in refinements block.\n"
+            << rParameters << std::endl;
+        KRATOS_ERROR_IF_NOT(rParameters.Has("geometry_type"))
+            << "Missing \"geometry_type\".\n"
+            << rParameters << std::endl;
+        if (rParameters["geometry_type"].GetString() == "NurbsSurface") {
```
Comment (manuelmessmer) [inline applications/IgaApplication/custom_modelers/refinement_modeler.cpp:70]: Do we wanna enter a new function here? Then I will also adjust the volume refinement to be called from here.

File: applications/IgaApplication/custom_modelers/refinement_modeler.cpp [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                    }
+
+                    PointerVector<NodeType> PointsRefined;
+                    Vector KnotsURefined;
+                    Vector WeightsRefined;
+
+                    NurbsSurfaceRefinementUtilities::KnotRefinementU(*(p_nurbs_surface.get()), knots_to_insert_u,
+                        PointsRefined, KnotsURefined, WeightsRefined);
+
+                    // Recreate nodes in model part to ensure correct assignment of dofs
```
Comment (manuelmessmer) [inline applications/IgaApplication/custom_modelers/refinement_modeler.cpp:96]: I still think it should be possible to just Add the nodes to the model part.

File: applications/IgaApplication/custom_modelers/refinement_modeler.cpp [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            rGeometryList.push_back(rModelPart.pGetGeometry(rParameters["brep_name"].GetString()));
+        }
+        if (rParameters.Has("brep_names")) {
+            for (SizeType i = 0; i < rParameters["brep_names"].size(); ++i) {
+                rGeometryList.push_back(rModelPart.pGetGeometry(rParameters["brep_names"][i].GetString()));
+            }
+        }
+
+        KRATOS_ERROR_IF(rGeometryList.size() == 0)
+            << "Empty geometry list. Either \"brep_id\", \"brep_ids\", \"brep_name\" or \"brep_names\" are the possible options." << std::endl;
```
Comment (manuelmessmer) [inline applications/IgaApplication/custom_modelers/refinement_modeler.cpp:172]: here I would then add volume_id. We then just have to make sure that the nurbs_geometry_modeler is called before the refinement modeler.

File: kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+// Project includes
+#include "nurbs_surface_refinement_utilities.h"
+
+namespace Kratos {
+
+    ///@name Operations
+    ///@{
+
+    void NurbsSurfaceRefinementUtilities::KnotRefinementU(
```
Comment (manuelmessmer) [inline kratos/geometries/nurbs_shape_function_utilities/nurbs_surface_refinement_utilities.cpp:23]: Will also rename my files to NurbsVolumeRefinement ;)

File: kratos/tests/cpp_tests/geometries/test_nurbs_surface.cpp [hunk 6]
```diff
@@ -151,6 +152,37 @@ namespace Testing {
             points, p, q, knot_u, knot_v);
     }
 
+    NurbsSurfaceGeometry<3, PointerVector<NodeType>> GenerateReferenceNodeSurfaceWithWeightVector() {
+        PointerVector<NodeType> points;
+
+        points.push_back(NodeType::Pointer(new NodeType(1, 0, 5, 0)));
```
Comment (manuelmessmer) [inline kratos/tests/cpp_tests/geometries/test_nurbs_surface.cpp]: Could use make_intrusive_ptr.

**Review by manuelmessmer — COMMENTED** [review]
@tteschemacher  nice job. I do like the interface. Once this is merged, I will adjust the volumes accordingly.

File: applications/IgaApplication/custom_modelers/refinement_modeler.cpp (hunk 1 above)
Comment (tteschemacher) [reply applications/IgaApplication/custom_modelers/refinement_modeler.cpp:74]: either for this one or for the brep geometries.

File: applications/IgaApplication/custom_modelers/refinement_modeler.cpp (hunk 2 above)
Comment (tteschemacher) [reply applications/IgaApplication/custom_modelers/refinement_modeler.cpp:70]: would make sense ;)

File: kratos/tests/cpp_tests/geometries/test_nurbs_surface.cpp (hunk 6 above)
Comment (tteschemacher) [reply kratos/tests/cpp_tests/geometries/test_nurbs_surface.cpp]: good point. this function isn't needed anymore anyways ;)

**Comment by manuelmessmer** [comment]
@tteschemacher nice! Looks good to me.

**Review by manuelmessmer — APPROVED** [review]

**Comment by tteschemacher** [comment]
Thanks a lot @manuelmessmer for approving. 

We'll edit this modeler in the future some more times....

**Pull request merged**